## HTTP-Request methods

Source: https://www.medienpalast.net/glossar/begriff/http-request/

HTTP (Hypertext Transfer Protocol Request) defines a set of request methods to indicate the purpose of the request and what is expected if the request is successful.

The requests come from Client, typically a web browser and would be sent to a Server, to request resources. These resources can be websites, images, videos or have another format, that is available over the internet. The communication between Client and Server takes place over the HTTP-Protocol.


## HTTP-Methods

The most common HTTP methods are:

- GET: to retrieve information from the server

- POST: to serve to transmit data to the server

- PUT: to update the data

- DELETE: the delete resources


### Components

The first fine contains:

- the URL of the requested resource
- the HTTP method
- the HTTP version

Following the first line are the HEADER, which provide additional information about the request, such as the type of client or the accepted formats of the returned data.

POST and PUT requests also include a BODY containing the data to be transmitted. This body can be in various formats such as JSON, XML, or form data. The body is particularly important because it contains the information the server needs to process to fulfill the request.

### Why are HTTP Requests important?

They aim to provide the desired information quickly and efficiently. If an HTTP request fails due to an error or an invalid request, this often has an immediate impact on users: an error message is displayed, and the user does not receive the requested page.

### Error Handling

Servers can return various status codes to indicate the state of a request. For example, a successful GET request returns a 200 status code, while a 404 error indicates that the requested resource could not be found. These status messages are important for developers to identify and resolve errors.

In [32]:
import requests

In [33]:
req = requests.get("https://api.github.com")

In [34]:
# check if retrieve information is successful

req.status_code

200

In [35]:
# the retrieved information in format byte

req.content

b'{"current_user_url":"https://api.github.com/user","current_user_authorizations_html_url":"https://github.com/settings/connections/applications{/client_id}","authorizations_url":"https://api.github.com/authorizations","code_search_url":"https://api.github.com/search/code?q={query}{&page,per_page,sort,order}","commit_search_url":"https://api.github.com/search/commits?q={query}{&page,per_page,sort,order}","emails_url":"https://api.github.com/user/emails","emojis_url":"https://api.github.com/emojis","events_url":"https://api.github.com/events","feeds_url":"https://api.github.com/feeds","followers_url":"https://api.github.com/user/followers","following_url":"https://api.github.com/user/following{/target}","gists_url":"https://api.github.com/gists{/gist_id}","hub_url":"https://api.github.com/hub","issue_search_url":"https://api.github.com/search/issues?q={query}{&page,per_page,sort,order}","issues_url":"https://api.github.com/issues","keys_url":"https://api.github.com/user/keys","label_sea

In [36]:
req.text

'{"current_user_url":"https://api.github.com/user","current_user_authorizations_html_url":"https://github.com/settings/connections/applications{/client_id}","authorizations_url":"https://api.github.com/authorizations","code_search_url":"https://api.github.com/search/code?q={query}{&page,per_page,sort,order}","commit_search_url":"https://api.github.com/search/commits?q={query}{&page,per_page,sort,order}","emails_url":"https://api.github.com/user/emails","emojis_url":"https://api.github.com/emojis","events_url":"https://api.github.com/events","feeds_url":"https://api.github.com/feeds","followers_url":"https://api.github.com/user/followers","following_url":"https://api.github.com/user/following{/target}","gists_url":"https://api.github.com/gists{/gist_id}","hub_url":"https://api.github.com/hub","issue_search_url":"https://api.github.com/search/issues?q={query}{&page,per_page,sort,order}","issues_url":"https://api.github.com/issues","keys_url":"https://api.github.com/user/keys","label_sear

In [37]:
# information in JSON-Format

req.json()

{'current_user_url': 'https://api.github.com/user',
 'current_user_authorizations_html_url': 'https://github.com/settings/connections/applications{/client_id}',
 'authorizations_url': 'https://api.github.com/authorizations',
 'code_search_url': 'https://api.github.com/search/code?q={query}{&page,per_page,sort,order}',
 'commit_search_url': 'https://api.github.com/search/commits?q={query}{&page,per_page,sort,order}',
 'emails_url': 'https://api.github.com/user/emails',
 'emojis_url': 'https://api.github.com/emojis',
 'events_url': 'https://api.github.com/events',
 'feeds_url': 'https://api.github.com/feeds',
 'followers_url': 'https://api.github.com/user/followers',
 'following_url': 'https://api.github.com/user/following{/target}',
 'gists_url': 'https://api.github.com/gists{/gist_id}',
 'hub_url': 'https://api.github.com/hub',
 'issue_search_url': 'https://api.github.com/search/issues?q={query}{&page,per_page,sort,order}',
 'issues_url': 'https://api.github.com/issues',
 'keys_url': '

In [38]:
req.headers

{'Date': 'Sun, 27 Sep 2026 15:01:34 GMT', 'Cache-Control': 'public, max-age=60, s-maxage=60', 'Vary': 'Accept,Accept-Encoding, Accept, X-Requested-With', 'ETag': '"4f825cc84e1c733059d46e76e6df9db557ae5254f9625dfe8e1b09499c449438"', 'x-github-api-version-selected': '2022-11-28', 'Access-Control-Expose-Headers': 'ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Used, X-RateLimit-Resource, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type, X-GitHub-SSO, X-GitHub-Request-Id, Deprecation, Sunset, Warning', 'Access-Control-Allow-Origin': '*', 'Strict-Transport-Security': 'max-age=31536000; includeSubdomains; preload', 'X-Frame-Options': 'deny', 'X-Content-Type-Options': 'nosniff', 'X-XSS-Protection': '0', 'Referrer-Policy': 'origin-when-cross-origin, strict-origin-when-cross-origin', 'Content-Security-Policy': "default-src 'none'", 'Server': 'github.com', 'Content-Type': 'application/json; ch

In [39]:
# try GET-method with parameters

params = {"q": "requests+language:python"}
response = requests.get("http://api.github.com/search/repositories", params=params)

In [40]:
# response.json()

In [41]:
response.json()['items'][0]

{'id': 33210074,
 'node_id': 'MDEwOlJlcG9zaXRvcnkzMzIxMDA3NA==',
 'name': 'secrules-language-evaluation',
 'full_name': 'SpiderLabs/secrules-language-evaluation',
 'private': False,
 'owner': {'login': 'SpiderLabs',
  'id': 508521,
  'node_id': 'MDEyOk9yZ2FuaXphdGlvbjUwODUyMQ==',
  'avatar_url': 'https://avatars.githubusercontent.com/u/508521?v=4',
  'gravatar_id': '',
  'url': 'https://api.github.com/users/SpiderLabs',
  'html_url': 'https://github.com/SpiderLabs',
  'followers_url': 'https://api.github.com/users/SpiderLabs/followers',
  'following_url': 'https://api.github.com/users/SpiderLabs/following{/other_user}',
  'gists_url': 'https://api.github.com/users/SpiderLabs/gists{/gist_id}',
  'starred_url': 'https://api.github.com/users/SpiderLabs/starred{/owner}{/repo}',
  'subscriptions_url': 'https://api.github.com/users/SpiderLabs/subscriptions',
  'organizations_url': 'https://api.github.com/users/SpiderLabs/orgs',
  'repos_url': 'https://api.github.com/users/SpiderLabs/repos',


In [42]:
# note: with POST-method, the link must be correct so that the data can be posted

req = requests.post('http://httpbin.org/post',data = {'post-method':'testfromme'})
req.json()

# data that is sent to server is stored in form

{'args': {},
 'data': '',
 'files': {},
 'form': {'post-method': 'testfromme'},
 'headers': {'Accept': '*/*',
  'Accept-Encoding': 'gzip, deflate, zstd',
  'Content-Length': '22',
  'Content-Type': 'application/x-www-form-urlencoded',
  'Host': 'httpbin.org',
  'User-Agent': 'python-requests/2.33.1',
  'X-Amzn-Trace-Id': 'Root=1-6ab92fda-56afd9e171b649706e5c4f9d'},
 'json': None,
 'origin': '89.246.102.198',
 'url': 'http://httpbin.org/post'}

In [43]:
my_json = {"post-method":"test-json"}
req = requests.post('http://httpbin.org/post',json = my_json)
req.json()

{'args': {},
 'data': '{"post-method": "test-json"}',
 'files': {},
 'form': {},
 'headers': {'Accept': '*/*',
  'Accept-Encoding': 'gzip, deflate, zstd',
  'Content-Length': '28',
  'Content-Type': 'application/json',
  'Host': 'httpbin.org',
  'User-Agent': 'python-requests/2.33.1',
  'X-Amzn-Trace-Id': 'Root=1-6ab92fda-03a5c8ef5404336a1c9b8227'},
 'json': {'post-method': 'test-json'},
 'origin': '89.246.102.198',
 'url': 'http://httpbin.org/post'}

In [44]:
with open("http_test.txt", "w") as file:
    file.write("http_test")

In [45]:
my_file = {"file": open("http_test.txt", "rb")}
req = requests.post("http://httpbin.org/post",files = my_file)
req.json()

{'args': {},
 'data': '',
 'files': {'file': 'http_test'},
 'form': {},
 'headers': {'Accept': '*/*',
  'Accept-Encoding': 'gzip, deflate, zstd',
  'Content-Length': '158',
  'Content-Type': 'multipart/form-data; boundary=95e1bf50b62e9b62da017edbc4433d6d',
  'Host': 'httpbin.org',
  'User-Agent': 'python-requests/2.33.1',
  'X-Amzn-Trace-Id': 'Root=1-6ab92fda-04757e1e0b3ebca952c5ff7c'},
 'json': None,
 'origin': '89.246.102.198',
 'url': 'http://httpbin.org/post'}

In [46]:
# put-method

req = requests.put("http://httpbin.org/put",data = {"put-method":"test-put-method"})
req.json()

{'args': {},
 'data': '',
 'files': {},
 'form': {'put-method': 'test-put-method'},
 'headers': {'Accept': '*/*',
  'Accept-Encoding': 'gzip, deflate, zstd',
  'Content-Length': '26',
  'Content-Type': 'application/x-www-form-urlencoded',
  'Host': 'httpbin.org',
  'User-Agent': 'python-requests/2.33.1',
  'X-Amzn-Trace-Id': 'Root=1-6ab92fdb-7f5da91744048f7c02ffd97f'},
 'json': None,
 'origin': '89.246.102.198',
 'url': 'http://httpbin.org/put'}

In [47]:
# delete-method
url = "http://httpbin.org/get"
req = requests.delete("http://httpbin.org/delete")
req = requests.delete(url)

In [49]:
from getpass import getpass
requests.get('http://api.github.com/user',auth = ('username',getpass()))

<Response [401]>

The HTTP error 401 Unauthorized indicates that a request was sent without valid authentication or that authentication failed. This means that the user did not provide the necessary credentials to access the requested resource.

### HTTP Status-Code

Source: https://restfulapi.net/http-status-codes/

HTTP defines these standard status codes that can be used to convey the results of a client’s request. The status codes are divided into five categories.

- 1xx: Informational – Communicates transfer protocol-level information.
- 2xx: Success – Indicates that the client’s request was accepted successfully.
- 3xx: Redirection – Indicates that the client must take some additional action in order to complete their request.
- 4xx: Client Error – This category of error status codes points the finger at clients.
- 5xx: Server Error – The server takes responsibility for these error status codes.



## Beautiful Soup

Beautiful Soup is a python library for pulling data out of HTML and XML files. It works with your favorite parser to provide idiomatic ways of navigating, searching, and modifying the parse tree.

In [50]:
from bs4 import BeautifulSoup

In [52]:
# Pass in a string

resp = requests.get("http://data.pr4e.org/romeo.txt")
html = resp.text
soup = BeautifulSoup(html, 'html.parser')
print(soup)

But soft what light through yonder window breaks
It is the east and Juliet is the sun
Arise fair sun and kill the envious moon
Who is already sick and pale with grief



In [54]:
# pass in an open filehandle

with open("index.html") as file:
    soup = BeautifulSoup(file, 'html.parser')
    print(soup)

# or

# print(soup.prettify())

<html>
<head>
<title>
         Hello, Data Engineer
      </title>
</head>
<body>
<p class="title">
          Hello
         <b>
         The Dormouse's story
         </b>
</p>
<p class="story">
         Once upon a time there were three little sisters; and their names were
         <a class="sister" href="http://example.com/elsie" id="link1">
         Elsie
         </a>
<a class="sister" href="http://example.com/lacie" id="link2">
         Lacie
         </a>
         and
         <a class="sister" href="http://example.com/tillie" id="link3">
         Tillie
         </a>
         and they lived at the bottom of a well.
      </p>
<p class="story">
</p>
</body>
</html>


## Objects components

- Tag

In [56]:
# get element by tag
tag = soup.title
print(tag)

<title>
         Hello, Data Engineer
      </title>


- Name

In [57]:
# get name of element
name = tag.name
print(name)

title


In [59]:
# change name of tag
tag.name = "new_title_by_me"
print(tag)

<new_title_by_me>
         Hello, Data Engineer
      </new_title_by_me>


- Attributes

In [62]:
# get value of attributes
p = soup("p", "title")
p.result

[<p class="title">
           Hello
          <b>
          The Dormouse's story
          </b>
 </p>]

In [68]:
# access dictionary directly as .attrs
attr = soup.a
#print(attr)
a_attr = attr.attrs
print(type(a_attr), a_attr)

<class 'bs4.element.AttributeDict'> {'class': ['sister'], 'href': 'http://example.com/elsie', 'id': 'link1'}


A tag's attribute can be added, removed and modified. Again, this is done by treating the tag as a dictionary

In [69]:
tag = soup.a
print(tag.attrs)

{'class': ['sister'], 'href': 'http://example.com/elsie', 'id': 'link1'}


In [70]:
tag["class"] = "amazon"
tag["href"] = "https://www.amazon.de/"
print(tag)

<a class="amazon" href="https://www.amazon.de/" id="link1">
         Elsie
         </a>


In [71]:
del tag["class"]
del tag["href"]
print(tag)

<a id="link1">
         Elsie
         </a>


### Navigating the tree

Tags may contain ``strings`` and other ``tags``. These elements are the tag’s ``children``. Beautiful Soup provides a lot of different attributes for navigating and iterating over a tag’s children.

- Navigating using tag names

Using a tag name as an attribute will give you only the first tag by that name:

In [72]:
# get head element

head = soup.head
print(head)

<head>
<new_title_by_me>
         Hello, Data Engineer
      </new_title_by_me>
</head>


In [74]:
# get title element
title = soup.new_title_by_me
print(title)

<new_title_by_me>
         Hello, Data Engineer
      </new_title_by_me>


In [75]:
# get tag b in body
b = soup.body.b
print(b)

<b>
         The Dormouse's story
         </b>


In [77]:
# get all the <a> tags
list_a_tags = soup.find_all("a")
print(list_a_tags.result)

[<a id="link1">
         Elsie
         </a>, <a class="sister" href="http://example.com/lacie" id="link2">
         Lacie
         </a>, <a class="sister" href="http://example.com/tillie" id="link3">
         Tillie
         </a>]


- **``.contents`` and ``.children``**

In [79]:
cont_html = soup.head.contents
print(len(cont_html))

3


In [80]:
for content in cont_html:
    print(content.name, "<--")

None <--
new_title_by_me <--
None <--


In [81]:
body_children = soup.head.children
print(body_children)
for child in body_children:
    print(child.name, '<----')

<generator object Tag.children.<locals>.<genexpr> at 0x0000012228A72C20>
None <----
new_title_by_me <----
None <----


In [82]:
for child in soup.head.descendants:
    print(type(child), child)

<class 'bs4.element.NavigableString'> 

<class 'bs4.element.Tag'> <new_title_by_me>
         Hello, Data Engineer
      </new_title_by_me>
<class 'bs4.element.NavigableString'> 
         Hello, Data Engineer
      
<class 'bs4.element.NavigableString'> 



In [83]:
print(len(list(soup.children)))
print(len(list(soup.descendants)))

1
32


- **``.strings`` and ``.stripped_strings``**

In [84]:
# Use the .strings generator:
for string in soup.body.strings:
    print(repr(string))

'\n'
'\n          Hello\n         '
"\n         The Dormouse's story\n         "
'\n'
'\n'
'\n         Once upon a time there were three little sisters; and their names were\n         '
'\n         Elsie\n         '
'\n'
'\n         Lacie\n         '
'\n         and\n         '
'\n         Tillie\n         '
'\n         and they lived at the bottom of a well.\n      '
'\n'
'\n'
'\n'


In [85]:
# remove white space by using the .stripped_strings generator

for string in soup.body.stripped_strings:
    print(repr(string))

'Hello'
"The Dormouse's story"
'Once upon a time there were three little sisters; and their names were'
'Elsie'
'Lacie'
'and'
'Tillie'
'and they lived at the bottom of a well.'


### Searching the tree

- a string

In [87]:
# Finds all the <b> tags in the document
soup.find_all('b').result

[<b>
          The Dormouse's story
          </b>]

- a regular expression

In [88]:
import re
for tag in soup.find_all(re.compile("^b")):
    print(tag.name)

body
b


- a list

In [90]:
soup.find_all(["a", "b"]).result

[<b>
          The Dormouse's story
          </b>,
 <a id="link1">
          Elsie
          </a>,
 <a class="sister" href="http://example.com/lacie" id="link2">
          Lacie
          </a>,
 <a class="sister" href="http://example.com/tillie" id="link3">
          Tillie
          </a>]

- a function

In [92]:
def has_class_but_no_id(tag):
    return tag.has_attr('class') and not tag.has_attr('id')
soup.find_all(has_class_but_no_id).result

[<p class="title">
           Hello
          <b>
          The Dormouse's story
          </b>
 </p>,
 <p class="story">
          Once upon a time there were three little sisters; and their names were
          <a id="link1">
          Elsie
          </a>
 <a class="sister" href="http://example.com/lacie" id="link2">
          Lacie
          </a>
          and
          <a class="sister" href="http://example.com/tillie" id="link3">
          Tillie
          </a>
          and they lived at the bottom of a well.
       </p>,
 <p class="story">
 </p>]

 ### ``.find_all()``

Signature: ``find_all(name, attrs, recursive, string, limit, **kwargs)``

In [94]:
soup.find_all("new_title_by_me").result

[<new_title_by_me>
          Hello, Data Engineer
       </new_title_by_me>]

In [95]:
soup.find_all("p", "title").result

[<p class="title">
           Hello
          <b>
          The Dormouse's story
          </b>
 </p>]

In [97]:
soup.find_all("a").result

[<a id="link1">
          Elsie
          </a>,
 <a class="sister" href="http://example.com/lacie" id="link2">
          Lacie
          </a>,
 <a class="sister" href="http://example.com/tillie" id="link3">
          Tillie
          </a>]

In [98]:
soup.find_all(id="link2").result

[<a class="sister" href="http://example.com/lacie" id="link2">
          Lacie
          </a>]

In [99]:
soup.find(string=re.compile("sisters"))

'\n         Once upon a time there were three little sisters; and their names were\n         '

### Searching by CSS class

In [101]:
soup.find_all("a", class_="sister").result

[<a class="sister" href="http://example.com/lacie" id="link2">
          Lacie
          </a>,
 <a class="sister" href="http://example.com/tillie" id="link3">
          Tillie
          </a>]